klippa/src/graph.rs

```diff
@@ -0,0 +1,515 @@
+            ordering_scratch: Vec::with_capacity(count),
+            data: s.data(),
+            parents_invalid: true,
+            positions_invalid: true,
+            distance_invalid: true,
+            successful: true,
+            ..Default::default()
+        };
+
+        this.num_roots_for_space.push(1);
+        for obj_idx in packed_obj_idxs.iter() {
+            let obj = s.get_obj(*obj_idx)?;
+            let v = Vertex::from_object(obj);
+            if !v.link_positions_valid(count) {
+                return None;
+            }
+            this.vertices.push(v);
+        }
+
+        let mut i = 0;
+        this.ordering.resize_with(count, || {
+            i += 1;
+            count - i
+        });
+        Some(this)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn in_error(&self) -> bool {
+        !self.successful
+    }
+
+    pub(crate) fn sort_shortest_distance(&mut self) {
+        if self.vertices.len() < 2 {
+            return;
+        }
+
+        self.positions_invalid = true;
+        self.update_distances();
+
+        let v_count = self.vertices.len();
+        let mut queue = PriorityQueue::with_capacity(v_count);
+        self.ordering_scratch.resize(v_count, 0);
+        let mut removed_edges = vec![0_usize; v_count];
+
+        self.update_parents();
+
+        queue.push((self.root().modified_distance(0), self.root_idx()));
+        let mut order = 1_u32;
+        let mut pos = 0;
+
+        let new_ordering = &mut self.ordering_scratch;
+        while let Some((_, next_id)) = queue.pop() {
+            if pos >= v_count {
+                return;
+            }
+            new_ordering[pos] = next_id;
+            pos += 1;
+
+            let next_obj = &self.vertices[next_id].obj;
+            for link in next_obj
+                .real_links
+                .values()
+                .chain(next_obj.virtual_links.values())
+            {
+                let child_idx = link.obj_idx();
+                removed_edges[child_idx] += 1;
+
+                let child_v = &self.vertices[child_idx];
+                if child_v.incoming_edges() == removed_edges[child_idx] {
+                    queue.push((child_v.modified_distance(order), child_idx));
+                    order += 1;
+                }
+            }
+        }
+
+        std::mem::swap(&mut self.ordering, new_ordering);
+        if pos != v_count {
+            self.set_err();
+            //TODO: add print_orphaned_nodes()
+        }
+    }
+
+    fn root(&self) -> &Vertex {
+        &self.vertices[self.root_idx()]
+    }
+
+    pub(crate) fn update_parents(&mut self) {
+        if !self.parents_invalid {
+            return;
+        }
+
+        for v in self.vertices.iter_mut() {
+            v.reset_parents();
+        }
+
+        let count = self.vertices.len();
+        let mut real_links_idxes = Vec::with_capacity(count);
+        let mut virtual_links_idxes = Vec::with_capacity(count);
```

Comment:
Is this temp storage needed because of borrow checker? You could get away with just having one, process the real links first, clear, then process the virtual links.

Reply:
seems we still need two temp vectors because we have to save all link obj_idxes first, then we're done with immutable borrows.

Reply:
sounds good.